### DISCUSSION TITLE
Rdkit Reaction - restrict to specific type of amines

### DISCUSSION THREAD

**Discussion by YoavShamir5** [rdkit/rdkit#8034] [FAQ]
Hello, I am looking for a way to restrict the following amide coupling reaction to secondary amines only, here is my input to AllChem.ReactionFromSmarts:

```
reaction_smarts = "[c:4][C:1](=[O:2])O.[Nh:3] >> [c:4][C:1](=[O:2])[Nh:3]"
```

Is there straightforward syntax to do modify the pattern so that only secondary amines will react? Thanks!

**Comment by bp-kelley** [comment]
You might want to look at this:

https://github.com/rdkit/rdkit/blob/master/Data/Functional_Group_Hierarchy.txt

it has a collection of recursive smarts patterns for matching a large range of functional groups.

A recursive smarts matches a single atom with the specified environment ( this is the recursive bit )

we also have a mechanism where you can tag an atom and automatically add the recursive pattern.

**Comment by YoavShamir5** [comment]
Thanks a lot for the quick reply, Brian!
I am actually familiar with recursive smarts patterns in rdkit, but I am not sure how to combine this with the reaction syntax. 
In my example, if I want to react with secondary amine only, the relevant recursive pattern is: [N;H1;D2;$(N(-[#6])-[#6]);!$(N-C=[O,N,S])], but directly using this in the reaction does not work:

reaction_smarts = "[c:4][C:1](=[O:2])O.[N;H1;D2;$(N(-[#6])-[#6]);!$(N-C=[O,N,S])] >> [c:4][C:1](=[O:2])[N;H1;D2;$(N(-[#6])-[#6]);!$(N-C=[O,N,S])]"

What's the correct syntax for combining the reaction pattern with this recursive pattern? Thanks!

**Reply by bp-kelley** [reply]
If you place your code in backticks "```" it will be a lot easier for us to read the code.  

```
[N;H1;D2;$(N(-[#6])-[#6]);!$(N-C=[O,N,S])]

reaction_smarts = "[c:4][C:1](=[O:2])O.[N;H1;D2;$(N(-[#6])-[#6]);!$(N-C=[O,N,S])] >> [c:4][C:1](=[O:2])[N;H1;D2;$(N(-[#6])-[#6]);!$(N-C=[O,N,S])]"
```

You might be able to click "edit" on this comment to see how to add the markup

**Reply by bp-kelley** [reply]
The recursive smarts should be a drop in for the Atom itself.

Simple example:
```
from rdkit import Chem
from rdkit.Chem import AllChem

amines = [
    ("aliphatic N", "[N:1]"), 
    ("primary", "[N;H2;D1;$(N-!@[#6]);!$(N-C=[O,N,S]):1]"),
    ("secondary", "[N;H1;D2;$(N(-[#6])-[#6]);!$(N-C=[O,N,S]):1]"),
    ("tertiary", "[N;H0;D3;$(N(-[#6])(-[#6])-[#6]);!$(N-C=[O,N,S]):1]"),
]

for name, smarts in amines:
    rxn = AllChem.ReactionFromSmarts(f"{smarts}>>[N:1]C")
    for n in primary, secondary, tertiary:
        rxns = rxn.RunReactants([n])
        for prods in rxns:
            for res in prods:
                print(name, Chem.MolToSmiles(n), "=>", Chem.MolToSmiles(res))
```

**Reply by bp-kelley** [reply]
Result:
```
aliphatic N CN => CNC
aliphatic N CNC => CN(C)C
aliphatic N CN(C)C => CN(C)(C)C
primary CN => CNC
secondary CNC => CN(C)C
tertiary CN(C)C => CN(C)(C)C
```

**Reply by bp-kelley** [reply]
note that the last one produces a hypervalent molecule :)

**Reply by bp-kelley** [reply]
In general, you do not want smarts on the product side.  Here I use atom maps to map the amine to the target.  This is the 

```
>>[N:1]C
```

in the reaction.
